client/asset/btc/btc.go

```diff
@@ -0,0 +1,1071 @@
+// specified unspent coin. A slice of pubkeys required to spend the
+// coin and a signature for each pubkey are returned.
+func (btc *ExchangeWallet) SignMessage(coin asset.Coin, msg dex.Bytes) (pubkeys, sigs []dex.Bytes, err error) {
+	output, err := btc.convertCoin(coin)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error converting coin: %v", err)
+	}
+	tx, err := btc.wallet.GetTransaction(output.txHash.String())
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, txDetails := range tx.Details {
+		if txDetails.Vout == output.vout &&
+			(txDetails.Category == TxCatReceive ||
+				txDetails.Category == TxCatGenerate) {
+
+			pk, sig, err := btc.wallet.SignMessage(txDetails.Address, msg)
+			if err != nil {
+				return nil, nil, fmt.Errorf("error signing message with pubkey for address %s: %v", txDetails.Address, err)
+			}
+			pubkeys = append(pubkeys, pk)
+			sigs = append(sigs, sig)
+		}
+	}
+	if len(pubkeys) == 0 {
+		return nil, nil, fmt.Errorf("no valid keys found.")
+	}
+	return pubkeys, sigs, nil
+}
+
+// AuditContract retrieves information about a swap contract on the blockchain.
+// AuditContract would be used to audit the counter-party's contract during a
+// swap.
+func (btc *ExchangeWallet) AuditContract(coinID dex.Bytes, contract dex.Bytes) (asset.AuditInfo, error) {
+	txHash, vout, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+	// Get the contracts P2SH address from the tx output's pubkey script.
+	txOut, err := btc.node.GetTxOut(txHash, vout, true)
+	if err != nil {
+		return nil, fmt.Errorf("error finding unspent contract: %v", err)
+	}
+	pkScript, err := hex.DecodeString(txOut.ScriptPubKey.Hex)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding pubkey script from hex '%s': %v",
+			txOut.ScriptPubKey.Hex, err)
+	}
+	// Check for standard P2SH.
+	scriptClass, addrs, numReq, err := txscript.ExtractPkScriptAddrs(pkScript, btc.chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting script addresses from '%x': %v", pkScript, err)
+	}
+	if scriptClass != txscript.ScriptHashTy {
+		return nil, fmt.Errorf("unexpected script class %d", scriptClass)
+	}
+	// These last two checks are probably overkill.
+	if numReq != 1 {
+		return nil, fmt.Errorf("unexpected number of signatures expected for P2SH script: %d", numReq)
+	}
+	if len(addrs) != 1 {
+		return nil, fmt.Errorf("unexpected number of addresses for P2SH script: %d", len(addrs))
+	}
+	// Compare the contract hash to the P2SH address.
+	contractHash := btcutil.Hash160(contract)
+	addr := addrs[0]
+	if !bytes.Equal(contractHash, addr.ScriptAddress()) {
+		return nil, fmt.Errorf("contract hash doesn't match script address. %x != %x",
+			contractHash, addr.ScriptAddress())
+	}
+	// Get the receiving address.
+	_, receiver, stamp, _, err := dexbtc.ExtractSwapDetails(contract, btc.chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting swap addresses: %v", err)
+	}
```

Comment:
Could do `ExtractSwapDetails` first to avoid an unnecessary RPC and other ops if we are supplied with an invalid atomic swap contract.